windows/CodePush.cs

```diff
@@ -0,0 +1,668 @@
+        }
+
+        public async Task<string> GetJavaScriptBundleFileAsync(string assetsBundleFileName)
+        {
+            this.assetsBundleFileName = assetsBundleFileName;
+            string binaryJsBundleUrl = ASSETS_BUNDLE_PREFIX + assetsBundleFileName;
+            long binaryResourcesModifiedTime = await GetBinaryResourcesModifiedTime();
+            StorageFile packageFile = await codePushPackage.GetCurrentPackageBundle(this.assetsBundleFileName);
+            if (packageFile == null)
+            {
+                // There has not been any downloaded updates.
+                CodePushUtils.logBundleUrl(binaryJsBundleUrl);
+                isRunningBinaryVersion = true;
+                return binaryJsBundleUrl;
+            }
+
+            JObject packageMetadata = await codePushPackage.GetCurrentPackage();
+            long? binaryModifiedDateDuringPackageInstall = null;
+            string binaryModifiedDateDuringPackageInstallString = (string)packageMetadata[BINARY_MODIFIED_TIME_KEY];
+            if (binaryModifiedDateDuringPackageInstallString != null)
+            {
+                binaryModifiedDateDuringPackageInstall = long.Parse(binaryModifiedDateDuringPackageInstallString);
+            }
+
+            string packageAppVersion = (string)packageMetadata["appVersion"];
+
+            // TODO: test configuration
+            if (binaryModifiedDateDuringPackageInstall != null &&
+                    binaryModifiedDateDuringPackageInstall == binaryResourcesModifiedTime &&
+                    (IsUsingTestConfiguration() || appVersion.Equals(packageAppVersion)))
+            {
+                CodePushUtils.logBundleUrl(packageFile.Path);
+                isRunningBinaryVersion = false;
+                return FILE_BUNDLE_PREFIX + packageFile.Path.Replace(ApplicationData.Current.LocalFolder.Path, "").Replace("\\", "/");
+            }
+            else
+            {
+                // The binary version is newer.
+                didUpdate = false;
+                if (!mainPage.UseDeveloperSupport || !appVersion.Equals(packageAppVersion))
+                {
+                    await ClearUpdates();
+                }
+
+                CodePushUtils.logBundleUrl(binaryJsBundleUrl);
+                isRunningBinaryVersion = true;
+                return binaryJsBundleUrl;
+            }
+        }
+
+        private ApplicationDataContainer GetCodePushSettings()
+        {
+            return ApplicationData.Current.LocalSettings.CreateContainer(CODE_PUSH_PREFERENCES, ApplicationDataCreateDisposition.Always);
+        }
+
+        private JArray GetFailedUpdates()
+        {
+            ApplicationDataContainer settings = GetCodePushSettings();
+            string failedUpdatesString = (string)settings.Values[FAILED_UPDATES_KEY];
+            if (failedUpdatesString == null)
+            {
+                return new JArray();
```

Comment:
it looks to me like all of these two get settings functions are very similar (getFailedUpdates/getPendingUpdate). get settings, read from settings and return default if failure, then return parsed object from settings and write default to settings if failure. could we make a helper function for them to use and also so that if we need to expand to additional settings we can do so easily?